### PR TITLE
[Design2-99] Design System - Modal Component

### PR DESCRIPTION
**Pull request by levadadenys**
### [Design2-99] Design System - Modal Component

- implemented component
- used `CustomDialog` as `Modal` base
- added tests
- added storybook
- component is now `Ready for Dev`

~~Once https://github.com/code-dot-org/code-dot-org/pull/62614 is merged, plan is to use `CustomDialog` as a base component for `Modal` (same way as it's done for `Dialog` component).~~

https://github.com/user-attachments/assets/5b7e7beb-e8af-4b46-a59a-d7ea2454ae7d

## Links
- jira ticket: [Design2-99](https://codedotorg.atlassian.net/browse/DESIGN2-99)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/componentLibrary/modal/Modal.tsx
Comment (kelbyhawn) [inline apps/src/componentLibrary/modal/Modal.tsx]: It's best to have empty alt text if the image is descriptive, which I assume this will be most of the time. Can the default be swapped with empty alt text instead? cc @hannahbergam as our accessibility expert
```suggestion
          {imageUrl && <img src={imageUrl} alt={imageAlt || ""} />}
```

**Review by kelbyhawn — COMMENTED** [review]
Looks great, but I have one accessibility suggestion.

**Review by sanchitmalhotra126 — COMMENTED** [review]
> Once https://github.com/code-dot-org/code-dot-org/pull/62614 is merged, plan is to use CustomDialog as a base component for Modal (same way as it's done for Dialog component).

Will this significantly change how this component is implemented? Should we hold off on reviewing until the CustomDialog is merged?

**Comment by levadadenys** [comment]
> Will this significantly change how this component is implemented? Should we hold off on reviewing until the CustomDialog is merged?

I don't think it's that big of a change since it's pretty straight forward. CustomDialog is already merged, so I already updated this PR to use CustomDialog as a base for Modal compoent.
Visually nothing has changed, checked.


https://github.com/user-attachments/assets/5f79792e-0a16-4d47-af36-29c62cbf043a

File: apps/src/componentLibrary/modal/Modal.tsx
Comment (levadadenys) [reply apps/src/componentLibrary/modal/Modal.tsx]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img

> The alt attribute holds a textual replacement for the image, which is mandatory and incredibly useful for accessibility — screen readers read the attribute value out to their users so they know what the image means. Alt text is also displayed on the page if the image can't be loaded for some reason: for example, network errors, content blocking, or link rot.

There're no plans on adding figcaption to modal right now. No matter what the content of the image will be, if user won't be able to see it - user won't be able to understand it's content since screen reader won't have where to take image description from.

Thus, imgAlt is required here.

Comment (drizco) [reply apps/src/componentLibrary/modal/Modal.tsx]: if the image is decorative rather than informative, having a default alt text is more disruptive to users of screen readers than leaving it empty. What does "modal-image" convey to the user, and what would the experience be like to hear that alongside the informative text in the modal?

https://www.w3.org/WAI/tutorials/images/

Comment (kelbyhawn) [reply apps/src/componentLibrary/modal/Modal.tsx]: +1 to what Ryan said above. Here's more from that resource:

> Decorative images don’t add information to the content of a page. For example, the information provided by the image might already be given using adjacent text, or the image might be included to make the website more visually attractive.
> 
> In these cases, a null (empty) alt text should be provided (alt="") so that they can be ignored by assistive technologies, such as screen readers. Text values for these types of images would add audible clutter to screen reader output or could distract users if the topic is different from that in adjacent text. Leaving out the alt attribute is also not an option because when it is not provided, some screen readers will announce the file name of the image instead.

**Review by sanchitmalhotra126 — APPROVED** [review]

File: apps/src/componentLibrary/modal/Modal.tsx
Comment (levadadenys) [reply apps/src/componentLibrary/modal/Modal.tsx]: > What does "modal-image" convey to the user,

I agree it's not the best alt, added it here just to make sure that there'll always be an alt for the image.

Comment (levadadenys) [reply apps/src/componentLibrary/modal/Modal.tsx]: Thanks for the context! Based on your suggestions and w3 article I agree to @kelbyhawn's proposal, adding `alt={imageAlt || ""}`

Comment (levadadenys) [reply apps/src/componentLibrary/modal/Modal.tsx]: Знімок екрана 2024-11-26 о 21 22 02
Here how it will be interpreted in html code

**Review by kelbyhawn — APPROVED** [review]
🎉

**Review by hannahbergam — APPROVED** [review]
Thank you for updating that alt text default!

**Pull request merged**
